File: 0463-island-perimeter/0463-island-perimeter.java
class Solution {
        
    private static int[] dx = {-1, 0, 1, 0};
    private static int[] dy = {0, 1, 0, -1};
    
    public int islandPerimeter(int[][] grid) {
        if(grid == null || grid.length == 0)
            return 0;
        
        int x = grid.length;
        int y = grid[0].length;

        int sum = 0;
        for(int i = 0; i < x; i++){
            for(int j = 0; j < y; j++){
                sum += dfs(i, j, grid);
            }            
        }
        
        
        return sum;
    }
    
    public int dfs(int x, int y, int[][] grid)
    {        
        if(grid[x][y] != 1){
            return 0;
        }
        
        int count = 4;        
        
        for(int i = 0; i < 4; i++){
            int nx = x + dx[i];
            int ny = y + dy[i];
            
            if(nx >= 0 && nx < grid.length && ny >= 0 && ny < grid[0].length){
                if(grid[nx][ny] != 0){
                    --count;
                }
            }
            
        }
        
        return count;
    }
    
}

//

File: Easy/0463-island-perimeter/0463-island-perimeter.js
/**
 * @param {number[][]} grid
 * @return {number}
 */

var islandPerimeter = function(grid) {
    if(grid === null || grid.length === 0){
        return 0;
    }
    
    let sum = 0;
    for(let x = 0; x < grid.length; x++){
        for(let y = 0; y < grid[0].length; y++){
            sum += checkLand(x, y, grid);            
        }
    }
    
    return sum;
};


const checkLand = function(x, y, grid){
    if(grid[x][y] != 1){
        return 0;
    }
    
    const dx = [-1, 0, 1, 0];
    const dy = [0, 1, 0, -1];
    let nx = 0;
    let ny = 0;
    let count = 4;
    for(let i = 0; i < 4; i++){
        nx = x + dx[i];
        ny = y + dy[i];        
        if(nx >= 0 && nx < grid.length && ny >= 0 && ny < grid[0].length){
            if(grid[nx][ny] != 0){
                --count;
            }
        }
    }
    
    return count;
}